phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/PhoenixIndexImportDirectReducer.java

```diff
@@ -40,15 +43,79 @@
+                    IndexRebuildRegionScanner.VerificationResult.getVerificationResult(hTable, ts);
+            context.getCounter(PhoenixIndexToolJobCounters.SCANNED_DATA_ROW_COUNT).
+                    setValue(verificationResult.getScannedDataRowCount());
+            context.getCounter(PhoenixIndexToolJobCounters.REBUILT_INDEX_ROW_COUNT).
+                    setValue(verificationResult.getRebuiltIndexRowCount());
+            if (verifyType == IndexTool.IndexVerifyType.ONLY || verifyType == IndexTool.IndexVerifyType.BEFORE ||
+                    verifyType == IndexTool.IndexVerifyType.BOTH) {
+                context.getCounter(PhoenixIndexToolJobCounters.BEFORE_REBUILD_VALID_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getBeforeRebuildValidIndexRowCount());
+                context.getCounter(PhoenixIndexToolJobCounters.BEFORE_REBUILD_EXPIRED_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getBeforeRebuildExpiredIndexRowCount());
+                context.getCounter(PhoenixIndexToolJobCounters.BEFORE_REBUILD_MISSING_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getBeforeRebuildMissingIndexRowCount());
+                context.getCounter(PhoenixIndexToolJobCounters.BEFORE_REBUILD_INVALID_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getBeforeRebuildInvalidIndexRowCount());
+            }
+            if (verifyType == IndexTool.IndexVerifyType.BOTH || verifyType == IndexTool.IndexVerifyType.AFTER) {
+                context.getCounter(PhoenixIndexToolJobCounters.AFTER_REBUILD_VALID_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getAfterRebuildValidIndexRowCount());
+                context.getCounter(PhoenixIndexToolJobCounters.AFTER_REBUILD_EXPIRED_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getAfterRebuildExpiredIndexRowCount());
+                context.getCounter(PhoenixIndexToolJobCounters.AFTER_REBUILD_MISSING_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getAfterRebuildMissingIndexRowCount());
+                context.getCounter(PhoenixIndexToolJobCounters.AFTER_REBUILD_INVALID_INDEX_ROW_COUNT).
+                        setValue(verificationResult.getAfterRebuildInvalidIndexRowCount());
+            }
+            if (verificationResult.isVerificationFailed(verifyType)) {
+                throw new IOException("Index verification failed! " + verificationResult);
+            }
+        } catch (Exception e) {
+            throw new IOException("Fail to get index verification result", e);
+        }
+    }
+
+    @Override
+    protected void reduce(ImmutableBytesWritable arg0, Iterable<IntWritable> arg1,
+                          Reducer<ImmutableBytesWritable, IntWritable, NullWritable, NullWritable>.Context context)
+            throws IOException, InterruptedException
+
+    {
+        synchronized (this) {
```

Comment:
can use an AtomicBoolean instead of a synchronized block.

Reply:
I will use it